docs/Debugging.md

```diff
@@ -0,0 +1,81 @@
+Debugging
+=========
+
+SGX-LKL provides a wrapper around gdb called `sgx-lkl-gdb`.
+This wrapper is compatible with IDEs and can be used instead of `gdb`.
+`sgx-lkl-gdb` automatically loads the SGX-LKL gdb plugin which
+ensures that debug symbols (if available) are loaded correctly. 
+When running in hardware mode, `sgx-lkl-gdb` uses the corresponding SGX debug
+instructions to read from and write to enclave memory.
+
+## Prerequisites
+
+### Open Enclave
+
+`sgx-lkl-gdb` relies on the Open Enclave gdb plugin.
+[Install Open Enclave](https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/install_oe_sdk-Ubuntu_18.04.md) if necessary before continuing.
+
+The instructions below assume that Open Enclave was installed in the default directory `/opt/openenclave`.
+If you used a different directory, run `. <OE_PREFIX>/share/openenclave/openenclaverc` first.
+
+### Build mode
+
+Debugging enclave code (SGX-LKL as well as the application code) is only possible in debug and non-release builds of SGX-LKL.
+
+### Encrypted disks
+
+TODO add note on how to handle situations when disk keys are retrieved from a remote service that would not release them to debug/non-release builds of SGX-LKL (either re-create disks unencrypted, or, if available locally, hard-code key in app config for local debugging)
+
+### Confidential Containers
+
+The instructions below assume that disk images and configuration are available outside of a Confidential Container. Debugging of Confidential Containers (via Docker) is not supported currently.
+
+## Option A: Debugging with an SGX-LKL installation
+
+The instructions below assume that SGX-LKL was installed in `/opt/sgx-lkl`.
+
+To debug an application, invoke `sgx-lkl-gdb` as follows:
+```
+/opt/sgx-lkl/bin/sgx-lkl-gdb --args /opt/sgx-lkl/bin/sgx-lkl-run-oe --hw-debug root_disk.img /app
+```
+See `sgx-lkl-run-oe --help` for how to run using host and app config files.
+
+## Option B: Debugging with an SGX-LKL source build
+
+This assumes you want to debug SGX-LKL directly from the source tree without installing it.
+
+To debug an application, invoke `sgx-lkl-gdb` as follows:
+```
+tools/gdb/sgx-lkl-gdb --args build/sgx-lkl-run-oe --hw-debug root_disk.img /app
+```
+See `sgx-lkl-run-oe --help` for how to run using host and app config files.
+
+## Debugging advice
+
+### Visual Studio Code
+
+Visual Studio Code allows to use a custom `gdb` command for debugging.
+To use `sgx-lkl-gdb`, add `"miDebuggerPath": "/opt/sgx-lkl/bin/sgx-lkl-gdb"` (change path if necessary) in your `.vscode/launch.json` file.
+Tip: Use the "Visual Studio Code Remote - SSH" extension to debug on a remote VM with SGX support.
+
+### Run mode
+
+Apart from obvious differences (security, performance, ...) there are subtle differences between SGX-LKL software mode (`--sw-debug`) and hardware mode (`--hw-debug`).
```

Comment:
It's probably simpler to say in `sw-debug` mode we simulate the SGX hardware.

Reply:
Changed, also added a concrete example of what isn't simulated currently.